### DISCUSSION TITLE
Request for integrating "¬" symbol

### DISCUSSION THREAD

**Discussion by mbaggnx** [porrey/Virtual-ZPL-Printer#13] [Ideas]
Hello there, would be interesting in the next update to include the symbol "**¬**" as part of the mapping, because there are systems that use that symbol instead of "**^**"

For example:

`¬XA

¬FX Top section with logo, name and address.
¬CF0,60
¬FO50,50¬GB100,100,100¬FS
¬FO75,75¬FR¬GB100,100,100¬FS
¬FO93,93¬GB40,40,40¬FS
¬FO220,50¬FDIntershipping, Inc.¬FS
¬CF0,30
¬FO220,115¬FD1000 Shipping Lane¬FS
¬FO220,155¬FDShelbyville TN 38102¬FS
¬FO220,195¬FDUnited States (USA)¬FS
¬FO50,250¬GB700,3,3¬FS`

So the program can manage to read both characters:

![image](https://user-images.githubusercontent.com/119416732/204520590-fc4e1d56-3f44-4064-a029-44a5b9ff081e.png)

**Comment by porrey** [comment]
The main functionality is driven by the web service which I do not control. I coul add a feature to allow character replacement. This would at least let you test an application using your ZPL without modification.

**Reply by mbaggnx** [reply]
That would be awesome, because depending on the country and the soft that people uses, sometimes you have diferent characters for the same simbol.

This will be cool if you select on some option to translate that basis character "^" to any other one, (mostly its used ¬, but here are other ones too).

Thanks a lot! I will be waiting for its release!!!

**Reply by porrey** [reply]
@mbaggnx, I created a branch named [**ZPL-Filter**](https://github.com/porrey/Virtual-ZPL-Printer/tree/ZPL-Filter) with this new feature. The installer can be downloaded from the installer folder. Please note this new version requires .NET 7. If you do not have the run-time installed yet use download both the **MSI** and the **setup.exe** and launch the installer using the setup.

[Installer Folder](https://github.com/porrey/Virtual-ZPL-Printer/tree/ZPL-Filter/Installer)

I will merge this branch once I have sufficient feedback.

**Reply by mbaggnx** [reply]
> @mbaggnx, I created a branch named [**ZPL-Filter**](https://github.com/porrey/Virtual-ZPL-Printer/tree/ZPL-Filter) with this new feature. The installer can be downloaded from the installer folder. Please note this new version requires .NET 7. If you do not have the run-time installed yet use download both the **MSI** and the **setup.exe** and launch the installer using the setup.
> 
> [Installer Folder](https://github.com/porrey/Virtual-ZPL-Printer/tree/ZPL-Filter/Installer)
> 
> I will merge this branch once I have sufficient feedback.

Hello! 

I am glad to tell you that is working fine the character substitution option

Also, I just found 2 more things, 

1- bug?
You cannot select and modify back after using the port of printer and started the virtual printer:
![image](https://user-images.githubusercontent.com/119416732/206661397-8e40f5d0-3ddf-4e6a-bdf4-b80104bb400b.png)

2- would be interesting to add manualy or by importing the printers or ports/ips of the printers:
in my case I have like 127.0.0.2, 127.0.0.3 and 192.168..... so I can select 1 or various printers, the option of "Any" does not works propperly. As far as I used it on the ZPL Chrome extension, I can manage to select the diferent ports, but not in the program, so I can change printer easily

<img width="645" alt="2022-12-09_09h33_27" src="https://user-images.githubusercontent.com/119416732/206661688-508607c5-8d81-4444-980f-3a62737e9e98.png">


So as far as I tested, it is working cool the substitution of the characters option

Do you think it is possible to add this changes at the ZPL chrome extension?

Thanks for your time and effort!

**Reply by porrey** [reply]
I fixed the first issue and pushed a new installer to the same branch. I am not sure what you are asking for as described in #2.

**Reply by mbaggnx** [reply]
> I fixed the first issue and pushed a new installer to the same branch. I am not sure what you are asking for as described in #2.

In nº 2 I mean that I am not able to select more than 1 host, you just can select 127.0.0.1 and 127.10.20.11

For example, in chrome extension, you can edit or change the config of the host:
![image](https://user-images.githubusercontent.com/119416732/206706361-cb17c251-479b-409b-b410-dc0aa1d9ebb9.png)

As in ZPL Program you can only choose those 2 and the option "Any" wich I don´t know for what it is.

In my case, I have 3 printers:

XXXPRINTER001 HOST 127.0.0.2
XXXPRINTER002 HOST 127.0.0.3
XXXPRINTER003 IP (LDP) 192.162.3.3

so the thing was to be able to select or change the host/ip where to send as in chrome extension


And in 2 also I ask if it is possible to add to zpl chrome extension the substitution character option.

**Reply by porrey** [reply]
I can add custom IP address.

**Reply by porrey** [reply]
Editable IP address has been added and merged in the main branch (version 2.2.0).

**Comment by mbaggnx** [comment]
Okay! Also I have detected that the virtulicer of ZPL does not works with LDP/LPR port:
![image](https://user-images.githubusercontent.com/119416732/206712470-c3966d02-0116-42af-a24d-7e7b8e8fdd88.png)

It only works on Standart TCP/IP Port

Tested and it queeps in the quee

Is it possible to activate this option?

**Reply by porrey** [reply]
I do not plan to support LPR.

**Reply by mbaggnx** [reply]
Oh... it is hard to enable it?
Most of the use of this is at power IBM systems, since most protocols on virtual printers works on LPR wich are activated on the own windows with the virtualization process

Basicly it is possible to engage the virtual printer into the system

Thanks for all your effort!

**Reply by mbaggnx** [reply]
I was thinking, if it is hard to go on LPR port, it is possible to enable this kind of port? -> I mean, to add that type of string so we can rout it to LPR, since when we create a LPR port we can make it goes over TCP/IP (unless I am wrong)

127.0.0.8:QUEEPRINT


![2022-12-12_12h37_34](https://user-images.githubusercontent.com/119416732/207036228-1d478ecf-3ba2-4bed-83d7-7d4ce5aa4375.png)
![2022-12-12_12h37_39](https://user-images.githubusercontent.com/119416732/207036248-615dd976-3b1a-4169-9cae-92227a663a53.png)
